### PR TITLE
Feature 2694 thresholds v6

### PR DESCRIPTION
**Pull request by mordak**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/2694) ticket: https://redmine.openinfosecfoundation.org/issues/2694

Describe changes:
- Accept by_rule and by_both in rule thresholds.
- Add TIMEVAL convenience macros.
- Refactor threshold application to separate threshold table storage from threshold calculation.
- Ensure the by_rule threshold table is initialized when needed.
- Add tests for threshold parsing and application.
- Update docs for by_rule and by_both thresholds in rules.

Continuation of #4747

### CODE REVIEW

File: src/detect-engine-threshold.c [hunk 1]
```diff
@@ -713,6 +658,31 @@ void ThresholdHashInit(DetectEngineCtx *de_ctx)
...
+void ThresholdHashRealloc(DetectEngineCtx *de_ctx)
+{
+    /* Return if we are already big enough */
+    uint32_t num = de_ctx->signum + 1;
+    if (num <= de_ctx->ths_ctx.th_size)
+        return;
+
+    void *ptmp = SCRealloc(de_ctx->ths_ctx.th_entry, num * sizeof(DetectThresholdEntry *));
+    if (ptmp == NULL) {
+        SCLogWarning(SC_ERR_MEM_ALLOC, "Error allocating memory for rule thresholds"
```
Comment (jlucovsky) [inline src/detect-engine-threshold.c]: Please use a value prefixed with `SC_WARN_` .... These are in `util-error.[hc]`

Comment (mordak) [reply src/detect-engine-threshold.c]: I actually went looking for an appropriate SC_WARN to use here, but none of them seem to apply to something like a memory allocation failure. The only ones I can find that apply to memory allocation problems are SC_ERR, and there are examples of SCLogWarning(SC_ERR_MEM_*) in conf.c, log-pcap.c, log-tlsstore.c, and util-buffer.c, so I thought this was fine.

Would you like a new SC_WARN_MEM_ALLOC in util-error.h?

Comment (victorjulien) [reply src/detect-engine-threshold.c]: No need for that. We're considering switching to a much simpler scheme soon too.

Comment (jlucovsky) [reply src/detect-engine-threshold.c]: No ... your pr looks good to me as is.

We'll address the discrepancy in a different issue.

**Comment by mordak** [comment]
> @mordak are there associated Suricata-Verify tests?

Sorry, no. I didn't know that was a thing until just now.

**Comment by victorjulien** [comment]
Merged in #4785, thanks Todd!

**Comment by victorjulien** [comment]
> > @mordak are there associated Suricata-Verify tests?
> 
> Sorry, no. I didn't know that was a thing until just now.

If you're interested in doing a few tests, it would be great. The existing tests can serve as examples.

**Pull request closed without merging**